tensorflow/python/grappler/auto_mixed_precision_test.py

```diff
@@ -0,0 +1,555 @@
+  if color == 'g': # Gray node
+    return math_ops.sqrt(input_tensor, name=name)
+  if color == 'c': # Clear node
+    return nn.relu(input_tensor, name=name)
+  if color == 'b': # Black node
+    return math_ops.log(input_tensor, name=name)
+  raise ValueError("Invalid node color: " + str(color))
+
+
+def _build_intertwined_loop_graph(inp_a_colors, inp_b_colors, body_a_colors,
+                                  body_b_colors, out_a_colors, out_b_colors):
+  """ Builds a test graph with intertwined loops """
+  a = _input([8, 8])
+  for i, color in enumerate(inp_a_colors):
+    a = _make_node_with_color(color, a, 'inputA_%i' % i)
+  b = _input([8, 8])
+  for i, color in enumerate(inp_b_colors):
+    b = _make_node_with_color(color, b, 'inputB_%i' % i)
+  def body_a(x):
+    for i, color in enumerate(body_a_colors):
+      x = _make_node_with_color(color, x, 'bodyA_%i' % i)
+    return x
+  def body_b(x):
+    for i, color in enumerate(body_b_colors):
+      x = _make_node_with_color(color, x, 'bodyB_%i' % i)
+    return x
+  a, b = _loop_vars_intertwined(a, b, body_a, body_b)[2:]
+  for i, color in enumerate(out_a_colors):
+    a = _make_node_with_color(color, a, 'outputA_%i' % i)
+  for i, color in enumerate(out_b_colors):
+    b = _make_node_with_color(color, b, 'outputB_%i' % i)
+  a = array_ops.identity(a)
+  b = array_ops.identity(b)
+  return a, b
+
+
+def _get_config(auto_mixed_precision=True):
+  """ Returns a config object with automatic mixed precision
+      activated if appropriate """
+  if auto_mixed_precision:
+    rewrite_config = rwcpb2.RewriterConfig(
+        auto_mixed_precision=rwcpb2.RewriterConfig.ON,
+        # do not remove duplicated nodes
+        arithmetic_optimization=rwcpb2.RewriterConfig.OFF)
+  else:
+    rewrite_config = rwcpb2.RewriterConfig(
+        auto_mixed_precision=rwcpb2.RewriterConfig.OFF,
+        # do not remove duplicated nodes
+        arithmetic_optimization=rwcpb2.RewriterConfig.OFF)
+  rewrite_config.min_graph_nodes = -1
+  graph_options = cpb2.GraphOptions(
+      rewrite_options=rewrite_config, build_cost_model=1)
+  config = cpb2.ConfigProto(graph_options=graph_options)
+  config.graph_options.optimizer_options.opt_level = -1
+  return config
+
+
+def _is_cast_to_fp16(node_name):
+  return node_name.endswith('-CastToFp16-AutoMixedPrecision')
+
+
+def _is_cast_to_fp32(node_name):
+  return node_name.endswith('-CastToFp32-AutoMixedPrecision')
+
+
+def _count_casts(nodes):
+  num_to_fp16 = 0
+  num_to_fp32 = 0
+  for node in nodes:
+    if _is_cast_to_fp16(node.name):
+      num_to_fp16 += 1
+    elif _is_cast_to_fp32(node.name):
+      num_to_fp32 += 1
+  return num_to_fp16, num_to_fp32
+
+
+def _build_node_map(nodes):
+  node_map = {}
+  for node in nodes:
+    node_map[node.name] = node
+  return node_map
+
+
+class AutoMixedPrecisionTest(test.TestCase):
+  """Tests the Grappler auto mixed precision optimizer."""
+  MIN_GPU_ARCH = (7, 0)
+
```

Comment:
Add a TODO to add eager tests with a tf.function.

One potential issue is that with V2 tf.cond (and i think tf.while_loop), they use tf.functions for the predicate and bodies instead of Switch and Merge ops. I think in eager, V2 cond is done by default. This will force the boundaries of tf.conds and tf.while_loops to be in float32.

Reply:
I've added a TODO.